MvvmCross/Core/MvxSetupSingleton.cs

```diff
@@ -61,57 +55,60 @@ public virtual void EnsureInitialized()
         {
             lock (LockObject)
             {
-                if (_initialized)
-                    return;
-
-                if (IsInitialisedTaskCompletionSource == null)
+                if (IsInitialisedTaskCompletionSource != null)
                 {
-                    IsInitialisedTaskCompletionSource = StartSetupInitialization();
+                    if (IsInitialisedTaskCompletionSource.Task.IsCompleted)
+                        return;
+
+                    // Only wait for the result here if nobody's monitoring the state
+                    if (_currentMonitor == null)
+                    {
+                        IsInitialisedTaskCompletionSource.Task.GetAwaiter().GetResult();
+                        MvxLog.Instance.Trace("EnsureInitialized has already been called so now waiting for completion");
+                    }
                 }
                 else
                 {
-                    MvxLog.Instance.Trace("EnsureInitialized has already been called so now waiting for completion");
+                    StartSetupInitialization();
```

Comment:
This beaks the behaviour of EnsureInitialized which should block until setup is completed. Currently this PR breaks more than it fixes.

Reply:
Sounds reasonable. We need to test PRs like #2713 or this one more before merging...